oak-run/src/main/java/org/apache/jackrabbit/oak/run/FrozenNodeRefsUsingIndexCommand.java

```diff
@@ -0,0 +1,159 @@
+import org.apache.jackrabbit.oak.api.Type;
+import org.apache.jackrabbit.oak.commons.PathUtils;
+import org.apache.jackrabbit.oak.plugins.memory.GenericPropertyState;
+import org.apache.jackrabbit.oak.run.Utils.NodeStoreOptions;
+import org.apache.jackrabbit.oak.run.commons.Command;
+import org.apache.jackrabbit.oak.spi.commit.CommitInfo;
+import org.apache.jackrabbit.oak.spi.commit.EmptyHook;
+import org.apache.jackrabbit.oak.spi.state.ChildNodeEntry;
+import org.apache.jackrabbit.oak.spi.state.NodeBuilder;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+import org.apache.jackrabbit.oak.spi.state.NodeStore;
+
+import com.google.common.io.Closer;
+
+/**
+ * Scans and lists all references to nt:frozenNode and returns an exit code of 1 if any are found (0 otherwise).
+ * <p/>
+ * This variant uses the /oak:index/reference by scanning through that list (only)
+ * and checking if any reference points to an nt:frozenNode (under /jcr:system/jcr:versionStorage
+ * at depth > 7).
+ * <p/>
+ * Example: 
+ * <pre>
+ * java -mx4g -jar oak-run-*.jar frozennoderefsusingindex mongodb://localhost/&lt;dbname&gt;
+ * </pre>
+ */
+public class FrozenNodeRefsUsingIndexCommand implements Command {
+
+    public static final String NAME = "frozennoderefsusingindex";
+
+    @Override
+    public void execute(String... args) throws Exception {
+        String help = NAME + " {<path>|<mongo-uri>|<jdbc-uri>} [options]";
+        Utils.NodeStoreOptions nopts = new Utils.NodeStoreOptions(help).parse(args);
+        if (hasNtFrozenNodeReferences(nopts)) {
+            System.err.println("FAILURE: References (in /oak:index/references) to nt:frozenNode found.");
+            System.exit(1);
+        } else {
+            System.out.println("SUCCESS: No references (in /oak:index/references) to nt:frozenNode found.");
+        }
+    }
+
+    private static boolean hasNtFrozenNodeReferences(NodeStoreOptions nopts) throws IOException {
+        Closer closer = Utils.createCloserWithShutdownHook();
+        try {
+            NodeStore store = Utils.bootstrapNodeStore(nopts, closer);
+            NodeState root = store.getRoot();
+
+            NodeBuilder b = root.builder();
+            NodeBuilder content = b.child("content");
+            PropertyState refProp = GenericPropertyState.referenceProperty("myref", "dd31e965-7e74-4c04-9a64-6698ecb4c624");
+            content.setProperty(refProp);
+            store.merge(b, EmptyHook.INSTANCE, CommitInfo.EMPTY);
+
+            NodeState oakIndex = root.getChildNode("oak:index");
+            NodeState refIndex = oakIndex.getChildNode("reference");
+            NodeState uuidIndex = oakIndex.getChildNode("uuid");
+            NodeState uuids = uuidIndex.getChildNode(":index");
+            System.out.println("Scanning ... refindex = " + refIndex);
+            NodeState references = refIndex.getChildNode(":references");
+
+            List<FrozenNodeRef> list = new LinkedList<FrozenNodeRef>();
```

Comment:
Same concern about memory as above.

Reply:
Agree - changed to a `count` at https://github.com/apache/jackrabbit-oak/pull/238/commits/fb8b330c4f248078bb2dce62cfb19ae90a50030c